### DISCUSSION TITLE
Linking error in Release mode

### DISCUSSION THREAD

**Discussion by rcarver** [pointfreeco/swift-composable-architecture#1556] [Q&A] [answered]
Hoping someone can explain what's going on here. In prepping some work for a Release build, I got some strange build errors. These only show up in a Release build. Debug works fine.

```
Undefined symbols for architecture arm64:
  "_$s18FirstLaunchFeature011WelcomeBackC0V4CoreV22ComposableArchitecture15ReducerProtocolAAWP", referenced from:
      _$s22ComposableArchitecture17_IfCaseLetReducerV6reduce4into6actionAA15EffectPublisherVy6ActionQzs5NeverOG5StateQzz_AJtFACy18FirstLaunchFeature011WelcomeBackQ0V4CoreVAP0R6ScreenVG_AP012UpgradeOfferU0VTgq5 in AppFeature.o
  "_$s18FirstLaunchFeature011WelcomeBackC0V4CoreVN", referenced from:
      _$s22ComposableArchitecture17_IfCaseLetReducerV6reduce4into6actionAA15EffectPublisherVy6ActionQzs5NeverOG5StateQzz_AJtFACy18FirstLaunchFeature011WelcomeBackQ0V4CoreVAP0R6ScreenVG_AP012UpgradeOfferU0VTgq5 in AppFeature.o
ld: symbol(s) not found for architecture arm64
```

Here's the failing type:

```swift
public struct WelcomeBackFeature: ReducerProtocol {
  public enum State: Equatable {
    case welcomeScreen(WelcomeScreen.State)
    case upgradeOfferScreen(UpgradeOfferScreen.State)
    init() { self = .welcomeScreen(.init()) }
  }
  public enum Action: Equatable {
    case welcomeScreen(WelcomeScreen.Action)
    case upgradeOfferScreen(UpgradeOfferScreen.Action)
  }
  public var body: some ReducerProtocol<State, Action> {
    Core()
      .ifCaseLet(/State.welcomeScreen, action: /Action.welcomeScreen, then: WelcomeScreen.init)
      .ifCaseLet(/State.upgradeOfferScreen, action: /Action.upgradeOfferScreen, then: UpgradeOfferScreen.init)
  }
  struct Core: ReducerProtocol {
    func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
      switch action {
      case .welcomeScreen(.welcomeMessageContinueButtonTappped):
        state = .upgradeOfferScreen(.init())
        return .none
      case .upgradeOfferScreen:
        return .none
      }
    }
  }
}
```

(Side note: I've been exploring this pattern of a nested `Core` reducer type as opposed to a `core` property and liking it)

---

After much trial and error, the fix is to make `WelcomeBackFeature.Core` public. But why?

```swift
  public struct Core: ReducerProtocol {
    public func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
```

I haven't yet figured out how to reproduce this in a new project, nor with a minimal new feature in the same project. Starting with this, where does the problem come from?

```swift
public struct CoreContainingFeature: ReducerProtocol {
  public struct State: Equatable {
  }
  public enum Action: Equatable {
  }
  public var body: some ReducerProtocol<State, Action> {
    Core()
  }
  struct Core: ReducerProtocol {
    func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
      return .none
    }
  }
}
```

**Comment by tgrapperon** [comment, marked as answer]
Hey @rcarver. This is a bug I've noticed a few months ago. Swift optimizer sometimes prunes non-public nested types that are only referenced in `@ResultBuilder`s functions, like `ReducerProtocol` and `SwiftUI`'s `body`s.

I didn't manage to reproduce it in a standalone project, and I forgot to report it.

It can bite you really easily with internal or private enums that you're using as identifiers. That's why I'm now declaring them as public and relocate them in the reducer itself, beside `State` and `Action`.

We should definitely find a repro and report it properly, as the issue will occur more and more given the way we organize features nowadays.

**Reply by tgrapperon** [reply]
It is possible that it affects only SwiftPM and not regular apps. @rcarver, is this feature hosted in some SwiftPM `Module`?

**Reply by rcarver** [reply]
ah! Thank you @tgrapperon. 

1. Yes, the feature is in an SPM module. 
2. Now I can reproduce by embedding `CoreContainingFeature` in `WelcomeBackFeature`. Previously I guess it was being pruned completely.
3. Strange that I use this pattern elsewhere and it's building correctly

Regardless, I guess the solution is to make these nested types public as well as you suggest.

**Reply by tgrapperon** [reply]
If you manage to reproduce reliably, it would still be nice to be able to extract a repro for submitting a bug to Swift. I'm afraid that a vague description will not be sufficient to give it enough traction to be fixed.

**Reply by saroar** [reply]
I notice clear project run again work sometime for my case I have too this issue

**Reply by mihaho** [reply]
I have reproduced this in my project as well, although I do not use SPM packages, but rather a regular app with frameworks for specific features. My setup is similar, I have a public feature reducer that references some internal ones (they are not nested inside of it). I had to mark some of the internal ones as public to archive the app successfully.

**Reply by rzulkoski** [reply]
I just got bit by this as well, not very fun to track down the cause. Thanks for this post!

**Reply by stephencelis** [reply]
Sadly we'll probably deprecate the type-based API as a result of this. Sorry you got bit by it, @rzulkoski!

**Reply by tgrapperon** [reply]
We can probably use a dependency to provide contextualized identifiers with a simple subscript:
```swift
@Dependency(\.id["fetchRequest"]) var fetchRequestID
```
If we wire this thing to `navigation-beta`'s `NavigationID`, we can even avoid cross-cancellation like it can happen in document apps.

**Reply by rzulkoski** [reply]
@stephencelis Sorry I wasn't more clear: I was bit by the compiler pruning an internal type, which was NOT an identifier. In my use case the user hands me an identified array of a public item type, which I transform into an identified array of an internal type. I don't have time at the moment to elaborate, but this post helped me pinpoint the issue, although the pruned type was a different use-case entirely.

**Reply by tgrapperon** [reply]
Was it only referenced in some `resultBuilder` decorated function/variable, like some `View`'s body or some `ReducerProtocol`'s body?

**Reply by rzulkoski** [reply]
It was a private enum used as an internal property of a public `ReducerProtocol`’s `State`. The fix was to make the enum public, and I was able to keep the property marked as internal.

**Reply by tgrapperon** [reply]
OK, thanks! I was thinking this was related to uses in `@resultBuilder` that weren't spotted and then pruned at the optimisation phase, but I'm not sure anymore after your report. Is this property accessed (only) from within a reducer's body, or are you touching it through some `ViewStore`'s observing `{ $0 }`?

**Comment by JimRoepcke** [comment]
Has anyone submitted a feedback on this yet? We've been bit by this problem as well.

**Reply by tgrapperon** [reply]
I didn't because I wasn't able to extract a repro, and I didn't want to submit a bug report without it.

**Comment by jgoodrick** [comment]
Just a heads up, I ran into exactly this due to an internal struct (that conformed to the ReducerProtocol) where its State and Action types were just internal typealiases of a parent reducer, as well. Making the struct and its State/Action public fixed it immediately.
